doc/dev/msgr2.rst

```diff
@@ -90,6 +102,8 @@ the form::
   [payload padding -- only present after stream auth phase]
   [signature -- only present after stream auth phase]
 
+* stream_id is generated by the client.
```

Comment:
Okay, I think we need to rethink this.  It makes sense for the client -> osd|mds|mon type connections, but not for mds <-> mds or osd <-> osd.  We want to open up a TCP connection between two peers, and establish streams between them for all fot he relevant entities, reusing that connection.  I think we need to make a distinction between client/server for the *transport* and the *stream*.  I think what I originally meant here is that in this context the client here is the stream initiator.  Imagine osd 0,1 in process A and 2,3 in proces B.  First osd.0 connects to osd.2, but later osd.2 or 3 needs to talk to osd.1, opening up a new stream within that same connection.

Reply:
perhaps we expand our terms at the top to include "stream initiator" and "stream target", and use those terms here?

Reply:
BTW I have some reservations about our terminology here because there is the Connection class in the code that represents and established stream.  I wonder if we should make the transport connection the 'transport stream' and use 'connection' for the logical exchange?

All of this is eventually going to bump into the weird/annoying racing connection reconciliation that is currently done in the messenger implementation for peer to peer connections (e.g., osd.0 and osd.1 sending racing intiail messages to each other).  FWIW I think mid- to long-term we will want to move that reconciliation into the OSD/MDS/mon, but maybe not yet.

Reply:
I agree, the socket connection, once established, should be re-used by both endpoints to create streams in any direction. I planned to do this but for some reason that I can't recall the implementation didn't follow this.

I also agree that we need to revise the terminology. Actually I am even in favor of changing the names of some classes in the messenger code. I don't like to call `Connection` to a stream. I would prefer to change the messenger interface and add a method called `getStream(endpoint)` and make the `Connection` class to refer to the actual socket connection. I know this would imply a major refactor of the code but I think it's worthwhile.